### PR TITLE
fix(TextEncoderStream): emoji encoding polyfill and add tests

### LINKED ISSUE #6458: [🐞] TextEncoderStream polyfill breaks resumability
### Which component is affected?

Qwik Runtime

### Describe the bug

In qwik-city 1.5.5 new [TextEncoderStream polyfill](https://github.com/QwikDev/qwik/pull/6310) was introduced. 

When deployed to Cloudflare useLexicalScope might return undefined for signals in some weird scenarios. Or qrls might stop working or anything else due to errors during resumability on client. 
I assume it occurs due to encoding issues, for example with new TextEncoderStream containers have `class="qc�"` when in 1.5.4 `class="qc📦"`. Some other symbols might also be corrupted, so qwik can't resume containers in edge cases.

Issue might not be reproducible if amount or sequence of counters is different or even if something irrelevant is removed like head nodes somewhere else.

### Reproduction

https://github.com/mhuretski/qwik-starter-1djsma

### Steps to reproduce

qwik-city 1.5.4 https://30efff55.check-qwik-155.pages.dev/ (OK)
qwik-city 1.5.5:
 - qwik 1.5.5 https://a2ac90cb.check-qwik-155.pages.dev/ (broken)
 - qwik 1.5.4 https://9fd9292e.check-qwik-155.pages.dev/ (broken)

1. click any of first 2 buttons
2. in 1.5.5 version containers are resumed corrupted when in 1.5.4 it's ok.

### System Info

```shell
cloudflare pages, qwik-city 1.5.5
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by octet-stream**
# Overview

Fixes #6458 

# What is it?

- [ ] Feature / enhancement
- [x] Bug
- [ ] Docs / tests / types / typos

# Description

The polyfill for TextEncoderStream introduced and proposed by me in https://github.com/QwikDev/qwik/pull/6310#issuecomment-2116414641 for use in Bun and Cloudflare runtimes breaks emoji encoding. This PR fixes the encoding issue.

The issue caused by how I iterated over the input string: for...of loop relies on iterator protocol, which treats an emoji as a single character and iterates through it just once, and then `String.charCodeAt(0)` call was picking up just half of the symbol. Meanwhile, the Node.js' implementation iterates over the string using for loop which treats emojis as two symbols and then encodes each part of that symbol.

# Use cases and why

Expected: The encoder stream polyfill should handle emoji symbols correctly.

Actual: The encoder stream polyfill breaks emoji symbols.

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/QwikDev/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by PatrickJS** [comment]
@octet-stream thanks a ton for looking into this and getting the test for it

**Review by PatrickJS — APPROVED** [review]

File: packages/docs/src/routes/api/qwik/index.md
Comment (PatrickJS) [inline packages/docs/src/routes/api/qwik/index.md:10650]: this is likely included because your fork might be outdated from main

Comment (octet-stream) [reply packages/docs/src/routes/api/qwik/index.md]: Ah, actually, I synced it, but forgot to pull changes to my local machine. Lemme fix this :)

Comment (PatrickJS) [reply packages/docs/src/routes/api/qwik/index.md]: I rebased it

Comment (PatrickJS) [reply packages/docs/src/routes/api/qwik/index.md]: no worries

Comment (octet-stream) [reply packages/docs/src/routes/api/qwik/index.md]: Or you can do this :)

File: packages/qwik-city/middleware/request-handler/polyfill.unit.ts
Comment (octet-stream) [inline packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: By the way, I found out that `.write` method returns a Promise, so it should be awaited in every other tests, as well as in this one. The tests are passing at their current state and are correct (except for one you testes sequential writing - I don't understand why are you comparing two promises returned by `.write method`), but I will send a PR with the improvement once I figure out why `.write` method freezes tests once when I await it. I'm not much familiar with WritableStream, so not sure what is wrong here.

Comment (PatrickJS) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: yeah I'm not sure either. the tests will retry x amount of times for anything async before giving up and timing out. do you want to make that test fix into another pr?

Comment (octet-stream) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: Yeah, this is just something to improve. The tests are still valid, as far as I can tell (they pass the expectations).

Comment (PatrickJS) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: ok lets ship this pr then

Comment (PatrickJS) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: well it looks like the e2e tests are hanging maybe because we need this in?

Comment (octet-stream) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: I don't see e2e hanging, but GitHub awaits for status report.

Am I looking to the right place?

Here https://github.com/QwikDev/qwik/actions/runs/9421547564/job/25958622059?pr=6466

and here https://github.com/QwikDev/qwik/actions/runs/9421547564/job/25958622214?pr=6466

Comment (PatrickJS) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: Screenshot 2024-06-07 at 1 23 09 PM

yeah, I'm not sure why it's not merging

Comment (octet-stream) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: Maybe the problem on GitHub actions side, this happens sometimes. I suggest we just wait, because this PR is not going anywhere :)

**Review by thejackshelton — APPROVED** [review]
PR Looks good to me! 

Thanks for adding tests 🙌 

Yeah let's wait for e2e tests to pass before we merge

**Review by gioboa — APPROVED** [review]
Thanks for your help 👏🚀

File: packages/qwik-city/middleware/request-handler/polyfill.unit.ts
Comment (shairez) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: fixed it 😊

Comment (octet-stream) [reply packages/qwik-city/middleware/request-handler/polyfill.unit.ts:82]: Thank you!

**Pull request merged**
